portfolio/src/main/java/com/google/sps/servlets/MemeHandler.java

```diff
@@ -0,0 +1,126 @@
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.FetchOptions;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.SortDirection;
+import com.google.appengine.api.images.ImagesService;
+import com.google.appengine.api.images.ImagesServiceFactory;
+import com.google.appengine.api.images.ServingUrlOptions;
+import com.google.gson.Gson;
+import com.google.sps.data.Meme;
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.Arrays;
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.List;
+import java.util.Map;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+
+@WebServlet("/meme-handler")
+public class MemeHandler extends HttpServlet {
+
+  /** Java object converter. */
+  private static final Gson gson = new Gson();
+
+  /** The max amount of memes that will be displayed. */
+  private int maxMemes = 40;
+
+  /** Read the data from the datastore and write it into /meme-handler as json. */
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    Query query = new Query("Meme").addSort("timeStamp", SortDirection.DESCENDING);
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery results = datastore.prepare(query);
+    // List<Entity> resultsList = results.asList(FetchOptions.Builder.withLimit(maxMemes));
+
+    List<Meme> memes = new ArrayList<>();
+    for (Entity entity : results.asIterable()) {
+        long id = entity.getKey().getId();
+        String author = (String) entity.getProperty("author");
+        String url = (String) entity.getProperty("url");
+        String description = (String) entity.getProperty("desc");
+        String timestamp = (String) entity.getProperty("timeStamp");
+
+        Meme memeObject = new Meme(id, author, url, description, timestamp);
+        memes.add(memeObject);
+    }
+
+    response.setContentType("application/json;");
+    String json = gson.toJson(memes);
+    response.getWriter().println(json);
+    }
+
+
+    /** Receive a meme and upload its data to the datastore. */
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    String description = request.getParameter("message");
+
+    String imageUrl = getUploadedFileUrl(request, "file");
+
+    Entity memeEntity = new Entity("Meme");
+    memeEntity.setProperty("author", "anonymous");
+    memeEntity.setProperty("url", imageUrl);
+    memeEntity.setProperty("desc", description);
+    memeEntity.setProperty("timeStamp", "11/22/11");
```

Comment:
looks like this shouldn't be hardcoded

Reply:
It shouldnt, gonna save incorporating users, and java.util.date in another pull request

Reply:
can you not set it? or at least add a todo comment?

Reply:
gotcha